transactions/update-contract/2023/dec-6-upgrade-dkg-staking-epoch/README.md

```diff
@@ -0,0 +1,98 @@
+# Update DKG, Staking, and Epoch Contracts
+
+> Dec 6th, 2023
+
+Upgrade to the DKG, Staking, and Epoch contracts from this pull request:
+
+- https://github.com/onflow/flow-core-contracts/pull/379
+
+These include performance changes, small additions to events, and the addition of the automatic access node slots. 
+
+## Transactions
+
+This upgrade has to be completed in multiple transactions:
+
+1. Upgrade `FlowDKG`.
+
+2. Upgrade `FlowIDTableStaking` and set access node slots.
+
+3. Upgrade `FlowEpoch`.
+
+Used this to generate the contract code arguments:
+
+`cat "./{Filename}.cdc" | xxd -p | tr -d '\n'`
+
+Verified using:
+```
+$ cat arguments-update-contract-{Filename}-mainnet.json | jq '.[1] | .value' | xxd -r -p > /tmp/temp.txt
+$ diff /tmp/temp.txt {Filename}.cdc
+(Should produce no difference)
+```
+___
+
+# Transaction 1
+
+## Transaction 1 Sequence of signing: 
+
+Signer: flow-staking
+Transaction: `templates/update_contract.cdc`
+Args: `arguments-update-FlowDKG.json`
+
+This transaction can be executed using the web tool.
+
+| Template                                                             | Arguments | Multisig Link   | Transaction |
+|----------------------------------------------------------------------|---        |---              |---          |
+| [update_contract.cdc](../../../../templates/update_contract.cdc) |  | | |
+
+
+### Results
+
+Successful attempt:
+https://flowscan.org/transaction/
+
+___
+
+# Transaction 2
+
+## Transaction 2 Sequence of signing: 
+
+Signer: flow-staking
+Transaction: `update_contract_set_access_slots.cdc`
+Args: `arguments-update-FlowIDTableStaking.json`
+
+This transaction can be executed using the web tool.
+
+| Template                                                             | Arguments | Multisig Link   | Transaction |
+|----------------------------------------------------------------------|---        |---              |---          |
+| [update_contract_set_access_slots.cdc](update_contract_set_access_slots.cdc) |  | | |
+
+
+### Results
+
+Successful attempt:
+https://flowscan.org/transaction/
+
+___
+
+# Transaction 3
+
+## Transaction 3 Sequence of signing: 
```

Comment:
We will also need to initialize the `/storage/flowEpochTimingConfig` path. 

I suggest we use the following config:
- `refTimestamp=1700074800` (Wed Nov 15 2023 19:00:00 GMT+0000 - the epoch transition that took place about 1 hour ago as I post this comment)
- `refCounter=101`
- `duration=604800` 1 week in seconds (24*7*60*60)

Modifying the transaction script to something like:
```cadence
transaction(name: String, code: String) {
	prepare(signer: AuthAccount) {
		signer.contracts.update__experimental(name: name, code: code.decodeHex())

		// Initialize the EpochTimingConfig in storage
		let config = FlowEpoch.EpochTimingConfig(
			duration: 604800, // 1 week in seconds
			refCounter: 101, // the epoch ended Wed Nov 15 @ 19:00UTC
			refTimestamp: 1700074800) // Unix timestampe for Wed Nov 15 @ 19:00UTC

		signer.save(config, to: /storage/flowEpochTimingConfig)
	}
}

Reply:
(By the way, we will need to set this config value for Canary/Testnet as well. I can help with that once we get there.)

Reply:
oh yeah, thank you for pointing that out. We'll actually need to do that in a separate transaction after we upgrade the `FlowEpoch` smart contract because the upgrade doesn't actually take effect until after the transaction is sealed, so we won't have access to the `EpochTimingConfig` type.

I pushed another [commit]([7124daa](https://github.com/onflow/service-account/pull/274/commits/7124daa358ec8260f8a232b0d4c198cf633ec28e)) with the new transaction and plan in the README